[RUNTIME] TCP Integration

brandonxxlee opened the pull request:
Adds TCP support for console logging and notifications with Dawn.

TCP process spawns once first message is received from Dawn and will send notifications and console logs via the notification proto over the tcp socekt.

Adds throttling to UDP receive process as well to hopefully improve performance.

Included compiled notification proto for now.

bodiddles committed:
[RUNTIME] Adds functionality to reset dawn's ip in runtime

Allows for resetting ip that runtime sends to avoid hard coding ip's
that runtime will send to.

When Dawn connects, its current ip address is stored into state manager and eventually sent to relevant processes.

Changes process spawning order to only spawn the UDP send process when Dawn connects. TCP process will also be spawned with this logic.

bodiddles committed:
Merge branch 'master' of https://github.com/pioneers/PieCentral into runtime/re_ip

bodiddles committed:
Merge branch 'master' of https://github.com/pioneers/PieCentral into runtime/re_ip

brandonxxlee committed:
Merge branch 'master' of github.com:pioneers/PieCentral into runtime/re_ip

bodiddles committed:
Merge branch 'runtime/re_ip' of https://github.com/pioneers/PieCentral into runtime/re_ip

brandonxxlee committed:
Merge branch 'runtime/re_ip' of github.com:pioneers/PieCentral into runtime/re_ip

brandonxxlee committed:
[RUNTIME] Lock IP on first connect to Dawn

brandonxxlee committed:
[RUNTIME] Start implementation of TCP

brandonxxlee committed:
[RUNTIME] Write main methods of TCP communication with Dawn

brandonxxlee committed:
[RUNTIME] Finish initial code, ready for testing

brandonxxlee committed:
[RUNTIME] Add method to terminate TCP if disconnected

brandonxxlee committed:
Merge branch 'master' of github.com:pioneers/PieCentral into runtime/TCPintegration

brandonxxlee committed:
[RUNTIME] Create ANSIBLE_COMMANDS enum class

brandonxxlee committed:
Merge branch 'master' of github.com:pioneers/PieCentral into runtime/TCPintegration

brandonxxlee committed:
[RUNTIME] Add Console Log command into runtimeUtil

bodiddles committed:
[RUNTIME] integrating tcp sockets

bodiddles committed:
Merge branch 'runtime/TCPintegration' of https://github.com/pioneers/PieCentral into runtime/TCPintegration

brandonxxlee committed:
[RUNTIME] Finish fake_dawn to implement TCP

bodiddles committed:
[RUNTIME] Enable TCP support

Adds TCP process and corresponding threads in Ansible.py.

TCP process creates TCP socket on port 1234 and starts two threads for sending and receiving
over the same socket.

Takes in commands from stateManager for handshake for studentCode upload and is able to send commands to Dawn.

Console logging is still untested

bodiddles committed:
Merge branch 'master' of https://github.com/pioneers/PieCentral into runtime/TCPintegration

bodiddles committed:
tcp sorta works

bodiddles committed:
Merge branch 'master' of https://github.com/pioneers/PieCentral into runtime/TCPintegration

bodiddles committed:
Merge branch 'master' of https://github.com/pioneers/PieCentral into runtime/TCPintegration

bodiddles committed:
deleted sending student code error to dawn

bodiddles committed:
[RUNTIME] TCP integration

Adds TCP support for console logging and notifications with Dawn.

TCP process spawns once first message is received from Dawn and will send notifications and console logs via the notification proto over the tcp socekt.

Adds throttling to UDP receive process as well to hopefully improve performance.

Included compiled notification proto for now.

bodiddles committed:
removed unnecessary print statements in ansible

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -5,6 +5,8 @@
 import runtime_pb2
 import ansible_pb2
+import notification_pb2
+
```
No need to add a newline here

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -223,15 +230,99 @@ def start(self):
         try:
             while True:
                 nextCall = time.time()
+                nextCall += 1.0/UDPRecvClass.UDP_RECV_HZ
```
You have a call to `select` right under this, which should be blocking. In that case, there should be absolutely no reason to have sleeps anywhere in this loop.

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -223,15 +230,99 @@ def start(self):
+        while True:
+            try:
+                rawMessage = pipe.recv()
+                nextCall = time.time()
```
Note that the sleeps in this thread appear to be for throttling purposes, rather than for handling use of non-blocking calls. There should be a comment indicating this.

nikitakit replied:
Upon further consideration, I don't think sleeps are the correct throttling mechanism here, either. But I'd be fine with merging this as-is, too, because it's not *too* bad.

Remember how you were debugging an issue where half of the message types were silently being dropped by the kernel? Perhaps that's actually the behavior we want -- your test code might have been spamming things at max speed, but the student code certainly won't be.

yizow replied:
The problem with throttling the sending liek this is that if whatever is on the other end of the `pipe` is generating data faster than this thread processes, its gonna fill up and cause a constant latency, as everything on the other end starts blocking on this function. 

Or, if we use non-blocking, the stuff on the other end will error out with `pipe full` errors. 
Especially since this is TCP, it doesn't seem like we want to drop any of these packets. 

It looks like this is purely used for 
1) confirming student code uploaded
2) sending over console output

The former is low frequency/ bandwidth and likely won't cause an issue.
The latter is potentially high frequency, high bandwidth, and most likely WILL cause an issue. Especially if you're throttling.

Maybe batching messages in the pipe together and sending them that way? 


In the interest of getting this merged though, I'm also okay with merging this.
But the console logging commit better fix it.

nikitakit reviewed (changes requested):
Just a few minor style comments.

Otherwise code LGTM, but it might be good to get another reviewer because I'm not very familiar with the networking part of the code.

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -223,15 +230,99 @@ def start(self):
                 self.unpackageData()
                 try:
-                    time.sleep(nextCall - time.time())
+                    time.sleep(nextCall +  - time.time())
```
So this line is wrong i think

But re: nikita's comments we shouldn't need this line at all

bodiddles committed:
Fixed style issues and sleep logic for throttling

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -222,16 +228,101 @@ def start(self):
         try:
             while True:
-                nextCall = time.time()
                 event = sel.select()
```
My previous comment regarding `select` remains unresolved.

bodiddles replied:
Could we use the sleeps for throttling reasons as we will be probably be receiving files faster than we want to be running this thread?

nikitakit replied:
This is a receiver, not a sender. Throttling belongs in sender (in this case, Dawn), not the receiver.

If you want to handle receiving packets faster than this thread can handle, the correct was to implement this is by dropping packets.

nikitakit replied:
Also, we went over this together during the fall semester. The way to implement rate-limited UDP receiver class is:

* Block until there is any data (using `select`)
* Read all available data from the kernel, dropping all but the last packet
* Process the single packet

Do you remember discussing this?

---

The reason sleeping is such a bad idea here is that it forces the kernel packet buffer to fill up, adding system latency equal to the kernel buffer size. If that happens you're going to end up with ridiculously large delays between moving a gamepad and the robot motors reacting. The whole reason we have a stateless UDP system is that any individual packet can be dropped without impacting the system as a whole.

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -151,6 +156,7 @@ def udpSender(self, badThingsQueue, stateQueue, pipe):
 
 class UDPRecvClass(AnsibleHandler):
     RECV_PORT = 1236
```
We should be using the global variables at the top of the file.

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -222,16 +228,101 @@ def start(self):
+            while True:
+                recv_data, addr = self.sock.recvfrom(2048)
+                unpackagedData = unpackage(recv_data)
+                if not unpackagedData:
```
EDIT: I misread this the first time.

Please check that `recvfrom` is actually a blocking call.

This condition of `if not unpackagedData` threw me off, and part of it is that unpackage currently doesn't ever return None. I think this condition should be removed, or an exception handler should be added to `unpackage`

bodiddles replied:
I changed the code to better reflect this.

bodiddles committed:
[Runtime] Changed throttling

Changed UDP Recv thread to not throttle.

Added code to TCP recv to better reflect its behavior.

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -150,15 +157,15 @@ def udpSender(self, badThingsQueue, stateQueue, pipe):
 class UDPRecvClass(AnsibleHandler):
-    RECV_PORT = 1236
-
+    UDP_RECV_HZ = 20
```
Delete this, you don't need `UDP_RECV_HZ` since you're not sleeping

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -150,15 +157,15 @@ def udpSender(self, badThingsQueue, stateQueue, pipe):
-    RECV_PORT = 1236
-
+    UDP_RECV_HZ = 20
+    
```
no need to make this whitespace change

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -5,15 +5,23 @@
+import notification_pb2
 from runtimeUtil import *
 
+SEND_PORT = 1235
```
Maybe name this `UDP_SEND_PORT`/`UDP_RECV_PORT` to distinguish from TCP now that TCP exists?

nikitakit reviewed (changes requested):
Just minor style comments this time.

Maybe @yizow should also review though, since I've had a bad track record of catching networking bugs.

bodiddles committed:
Removed UDP RECV_HZ and fixed whitespace

bodiddles committed:
minor text changes for udp ports

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -222,16 +227,90 @@ def start(self):
+                badThingsQueue.put(BadThing(sys.exc_info(), 
+                    "TCP receiver crashed with error: " + str(e),
+                    event = BAD_EVENTS.TCP_ERROR,
+                    printStackTrace = True))
```
You should keep the newline at eof

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -69,10 +77,9 @@ def start(self):
     packagerHZ = 20.0
     socketHZ = 20.0
-
+    
```
gratuitous whitespace change

nikitakit reviewed (approved):
lgtm now, ideally @yizow would also give this a look

bodiddles committed:
Fixed more whitespace errors
